pkg/engine/planner/logical/node_literal.go

```diff
@@ -0,0 +1,136 @@
+package logical
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/grafana/loki/v3/pkg/engine/planner/schema"
+)
+
+// LiteralKind denotes the kind of [Literal] value.
+type LiteralKind int
+
+// Recognized values of [LiteralKind].
+const (
+	// LiteralKindInvalid indicates an invalid literal value.
+	LiteralKindInvalid LiteralKind = iota
+
+	LiteralKindString    // String literal value.
+	LiteralKindInt64     // 64-bit integer literal value.
+	LiteralKindUint64    // 64-bit unsigned integer literal value.
+	LiteralKindByteArray // Byte array literal value.
+)
+
+var literalKindStrings = map[LiteralKind]string{
+	LiteralKindInvalid: "invalid",
+
+	LiteralKindString:    "string",
+	LiteralKindInt64:     "int64",
+	LiteralKindUint64:    "uint64",
+	LiteralKindByteArray: "[]byte",
+}
+
+// String returns the string representation of the LiteralKind.
+func (k LiteralKind) String() string {
+	if s, ok := literalKindStrings[k]; ok {
+		return s
+	}
+	return fmt.Sprintf("LiteralKind(%d)", k)
+}
+
+// A Literal represents a literal value known at plan time. Literal only
+// implements [Value].
+type Literal struct {
+	any any
+}
+
+var _ Value = (*Literal)(nil)
+
+// LiteralString creates a new Literal value from a string.
+func LiteralString(v string) *Literal { return &Literal{any: v} }
+
+// LiteralInt64 creates a new Literal value from a 64-bit integer.
+func LiteralInt64(v int64) *Literal { return &Literal{any: v} }
+
+// LiteralUint64 creates a new Literal value from a 64-bit unsigned integer.
+func LiteralUint64(v uint64) *Literal { return &Literal{any: v} }
+
+// LiteralByteArray creates a new Literal value from a byte slice.
+func LiteralByteArray(v []byte) *Literal { return &Literal{any: v} }
+
+// Kind returns the kind of value represented by the literal.
+func (lit Literal) Kind() LiteralKind {
+	switch lit.any.(type) {
```

Comment:
Should we need to check for `lit.any==nil`? Or is it supposed to be `LiteralKindInvalid`?

Reply:
Hm, I'll add support for a LiteralKindNull. Even if we don't have a way of representing that in LogQL right now, it's probably a useful semantic to have in the Literal.